File: backend/src/main/scala/domains/problem/model/ProblemDataManifest.scala
package domains.problem.model

final case class ProblemDataManifestEntry(
  path: ProblemDataPath,
  sizeBytes: Long,
  sha256: String
)

final case class ProblemDataManifest(
  problemSlug: ProblemSlug,
  entries: List[ProblemDataManifestEntry],
  version: String
)

object ProblemDataManifest:
  def fromEntries(problemSlug: ProblemSlug, entries: List[ProblemDataManifestEntry]): ProblemDataManifest =
    val sortedEntries = entries.sortBy(_.path.value)
    ProblemDataManifest(problemSlug = problemSlug, entries = sortedEntries, version = manifestVersion(sortedEntries))

  private def manifestVersion(entries: List[ProblemDataManifestEntry]): String =
    sha256Hex(
      entries
        .map(entry => s"${entry.path.value}:${entry.sizeBytes}:${entry.sha256}")
        .mkString("\n")
        .getBytes(java.nio.charset.StandardCharsets.UTF_8)
    )

  private def sha256Hex(bytes: Array[Byte]): String =
    java.security.MessageDigest
      .getInstance("SHA-256")
      .digest(bytes)
      .map("%02x".format(_))
      .mkString

File: frontend/src/objects/problem/ProblemDataManifest.ts
import type { ProblemSlug } from '@/objects/problem/ProblemSlug'
import type { ProblemDataManifestEntry } from '@/objects/problem/ProblemDataManifestEntry'

/** 判题侧题目数据清单；描述可下载数据文件、版本和所属题目。 */
export type ProblemDataManifest = {
  problemSlug: ProblemSlug
  entries: ProblemDataManifestEntry[]
  version: string
}
